### ISSUE TITLE
Trying to fine-tune BERT with skorch the the training procedure is wrong as the accuracy on the validation test sets is zero and the validation loss increases.

### ISSUE THREAD

**Issue by TLampr** [skorch-dev/skorch#625]
Hello,

I am trying to fine tune BERT in a multiclass setup on a dataset using skorch. I have this issue that my validation loss is increasing per epoch while the validation accuracy is constant and equal to zero. The train loss seems to be decreasing but it is not working correctly as after the training, when testing the train accuracy on the fitted model it is also approximately zero. My dataset is a list of strings. I am using a pipeline in which I have a step to tokenize the dataset according to the requirements of BERT and then I add my custom classifier. My classes consist of the indices of each class. I would be very grateful for any advice regarding on how to solve this issue. My transformation step is 

```
class Tokenize_transformer(BaseEstimator, TransformerMixin):
    def __init__(self, config):
        super(Tokenize_transformer, self).__init__()
        self.tokenizer = AutoTokenizer.from_pretrained(config)

    def _tokenize(self, X, y=None):
        tokenized = self.tokenizer.encode_plus(X, max_length=20, add_special_tokens=True, pad_to_max_length=True)
        tokenized_text = tokenized['input_ids']
        return np.array(tokenized_text)


    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
        X = np.array([self._tokenize(string) for string in tqdm(X)])
        return X

    def fit_transform(self, X, y=None, **fit_params):
        self = self.fit(X, y)
        return self.transform(X, y)
```

and my custom network is

```
class Transformer(nn.Module):
    def __init__(self, config, num_labels=213, dropout_proba=.1):
        super(Transformer, self).__init__()
        self.num_labels = num_labels
        self.model = AutoModel.from_pretrained(config)
        self.dropout = torch.nn.Dropout(dropout_proba)
        self.classifier = torch.nn.Linear(768, num_labels)


    def forward(self, X, **kwargs):
        attention_mask = np.zeros(X.numpy().shape)
        attention_mask[np.where(X.numpy() != 0)[0], np.where(X.numpy() != 0)[1]] = 1
        attention_mask = torch.tensor(attention_mask)
        _, X = self.model(X, attention_mask)
        X = self.dropout(X)
        X = self.classifier(X)
        return X
```

and I am initialising the net work as

```
  transformer = Transformer(config)
  net = NeuralNetClassifier(transformer, lr=2e-5, max_epochs=2, criterion=torch.nn.CrossEntropyLoss, optimizer=AdamW, callbacks=[ProgressBar(postfix_keys=['train_loss', 'valid_loss'])], train_split=CVSplit(cv=2, random_state=0))
```

and fitting the pipeline in this manner

`pipeline.fit(X=dataset.training_samples, y=dataset.training_labels)`

I have tried to follow the examples given in the documentation as close as possible. Thank you in advance for your time and any comments on how to fix this.

**Comment by BenjaminBossan** [comment]
Hi @TLampr,

At first glance, I don't spot any obvious mistake. However, it's a bit hard to read your code, could you format it correctly please (use three backticks at the start and end:  ```).

Furthermore, it would help a lot if you could attach a copy or picture of your train log and the code (and associated values) to generate your train and valid scores.

Just a thought: Could it be that your model is simply overfitting on your training data and that explains the result? Do you have any indication that the model should reach better validation scores (e.g. because you trained it without skorch and it did much better there)?

**Comment by TLampr** [comment]
Hello @BenjaminBossan ,

Thank you for your reply. I am sorry for the bad initial formatting. Here is a screen shot of a training session of 10 epochs. I am quite sure the problem is not overfitting because when I checked the accuracy of the network on my train set (to see if the problem is on the validation set) it was also zero. The model is provided pretrained and it is supposed to be finetunned for very few epochs with a very small learning rate which the developers suggest. The fine-tune that I am trying to do is supposed to overfit at about 3-4 epochs but here I am doing something wrong clearly because the validation metrics should not behave like that. 
<img width="438" alt="Screenshot 2020-04-22 at 16 05 32 1" src="https://user-images.githubusercontent.com/43912315/80277778-b6c7a900-86f1-11ea-90aa-d0cab47c9636.png">

Also the way I initialise my pipeline is

```
pipeline = Pipeline(
        [
            ("tokenizer", Tokenize_transformer(config)),

            ("classifier", _get_new_transformer(config, options.hyper_parameters))
        ]
)
```

If you need more information I would be very happy to provide it. Thank you for taking the time to look at this issue.

**Comment by BenjaminBossan** [comment]
Thanks for formatting the code. I think it looks good. Some minor things could be improved, but they're not related to your problem. When it comes to the attention mask, I can't comment without further context if it's correct but I trust you know what you're doing there.

> The model is provided pretrained and it is supposed to be finetunned for very few epochs with a very small learning rate which the developers suggest.

This sounds to me like you're working according to some tutorial, could you post the link to that if it's available?

> I am quite sure the problem is not overfitting because when I checked the accuracy of the network on my train set (to see if the problem is on the validation set) it was also zero.

So I made a quick check, with 213 labels, if each class were equally likely, the loss should be 5.36 if the model guesses randomly. That is roughly what you see after the first epoch for both train and valid data. After that, train loss improves considerably, so I agree that an accuracy of 0 should not happen.

Even if the model didn't learn anything, you would expect that just by accident, some samples should be predicted correctly. To investigate further, could you maybe paste the result from `pipeline.predict_proba(X_test)` and `pipeline.predict(X_test)`, as well as `y_test`, where `X_test` and `y_test` are 20 to 100 samples or so from your data (after the model has trained for a few epochs)?

Also, if you didn't already, could you just try out different hyper-parameters and see if anything helps (e.g. different optimizer, learning rate, dropout rate). I would recommend to use much less data so that you can iterate faster. Moreover, if you're not already doing that, please fix the parameters from `AutoModel` so that you only train the `classifier` part.

**Comment by TLampr** [comment]
Here are the results a `pipeline.predict_proba()`  
<img width="961" alt="Screenshot 2020-04-26 at 12 13 46" src="https://user-images.githubusercontent.com/43912315/80304862-32d8f400-87b9-11ea-8712-6bf926be288f.png">
<img width="964" alt="Screenshot 2020-04-26 at 12 14 57" src="https://user-images.githubusercontent.com/43912315/80304874-5c921b00-87b9-11ea-97b8-0754e370bf34.png">
<img width="475" alt="Screenshot 2020-04-26 at 12 18 07" src="https://user-images.githubusercontent.com/43912315/80304897-82b7bb00-87b9-11ea-9e39-9b1e222c2e5a.png">
and a `pipeline.predict()` on a small sample of the test set
<img width="472" alt="Screenshot 2020-04-26 at 12 17 15" src="https://user-images.githubusercontent.com/43912315/80304892-7d5a7080-87b9-11ea-8e4e-89119045f0ff.png">
In the `pipeline.predict_proba()` there values are practically identical and also there are present negative values and values bigger than 1.
In the `pipeline.predict()` the result always is label 163.
These results were generated after the model trained for 4 epochs.

Fine-tuning `self.model()`is also part of my experiments and it is left with unfrozen weights on purpose. Still I tried freezing its weights and fine-tune only the final fully connected layer (referred as `self.classifier`) but the results were similar. I have tried to use a different optimiser, specifically `torch.optim.Adam`but the results were the same. The hyper-parameters used are the ones the developers of the pre-trained BERT model (the one referred to the code as `AutoModel.from_pretrained(config)`) have suggested. Initially the `attention_mask` was generated automatically along with the tokenisation of X, returning a list of tuples each containing the tokens of a sample and its corresponding mask. The results were the same and I decided to change that, in the case that returning a list of tuples was the reason I had this issue. I use to model and tutorials from the transformer library https://huggingface.co which is a library that contains pre-trained NLP networks such as BERT, ALBERT, XLNet and more. The documentation for the network I use is https://huggingface.co/transformers/model_doc/bert.html and this is the git page https://github.com/huggingface/transformers of the library that includes instruction on how to fine-tune the networks.

**Comment by BenjaminBossan** [comment]
> there are present negative values and values bigger than 1.

This is because you don't use a nonlinearity for your final output. That should be okay, however, since you use `CrossEntropyLoss`. Still, you could try `NLLLoss` in conjunction with `softmax` nonlinearity on the output. At least that should give you well formed probabilities from `predict_proba`.

Also, just as a good practice, you should add `iterator_train__shuffle=True` to the net's init parameters (except if you have a good reason not to).

> In the `pipeline.predict()` the result always is label 163.

This is a typical problem: when the model doesn't train at all, it will just tend to predict the same class for everything. Then, if you don't have that label in your test set, accuracy will obviously be 0.

> Still I tried freezing its weights and fine-tune only the final fully connected layer (referred as `self.classifier`) but the results were similar.

This is surprising. At least each epoch should run much faster, making experimentation easier. **In the phase you're in, it's crucial that you can iterate fast, so I would keep on freezing the BERT model until you start seeing improvements.**

Regarding your data, are you using some kind of public dataset or some private/company data? How many samples do you have? How well balanced are the classes (e.g. frequency of the most/least frequent class).

> The hyper-parameters used are the ones the developers of the pre-trained BERT model (the one referred to the code as `AutoModel.from_pretrained(config)`) have suggested

Can you point me to where exactly? I saw [this section](https://github.com/huggingface/transformers#optimizers-bertadam--openaiadam-are-now-adamw-schedules-are-standard-pytorch-schedules) but there the hyper-parameters are different.

**Comment by TLampr** [comment]
Hi,

I would like to thank you very much for the time you invest in my problem.
I removed the ```iterator_train__shuffle=True``` before I posted the code here as I was trying to figure out what might be going wrong, thinking it might be mixing the labels somehow. The dataset I am using is private. It contains 90k samples and is highly imbalanced with the most frequent classes having about 1000 samples and the least frequent about 20. It is part of my experiments to evaluate different methods on that dataset and one of them is BERT, pretrained and fine-turned. I will freeze the pretrained model again and experiment on parameters as you suggest.

As I am only fine-running the network on my dataset and I don't want to retrain it, I am using some of the parameters suggested in the fine-tune sections of this guide for which I post links bellow.

 https://github.com/huggingface/transformers#run_gluepy-fine-tuning-on-glue-tasks-for-sequence-classification

https://github.com/huggingface/transformers#fine-tuning-bert-model-on-the-mrpc-classification-task

I have already used BERT as a transformer in a pipeline with SGDclassifier as part of my experiments with BERT pre-trained and my results were at the range of 75% f1-score macro. 
Since there doesn't seem to be a problem of my code formulation, and as you suggested that I should use different parameters, do you have suggestions about what I should change? Thank you again for your time.

**Comment by BenjaminBossan** [comment]
Hmm, it's really hard to remotely debug a net that's not learning. I can only make a few general suggestions:

* did you try changing the loss function as I suggested above?
* try larger and smaller batch sizes
* try different optimizers, start with a simple `SGD`
* try to add an additional linear layer between the BERT output and the classification layer (with relu)
* turn off dropout completely

While doing all those experiments, you should freeze the BERT weights and use much less data to achieve quick iteration. That should at least lead to a good train loss. Once you've got that, try to increase the number of samples to get better valid scores, tuning the hyper-parameters as needed.

Another experiment could be to filter the data to, say, only the two most frequent classes, thus making it a much simpler problem. If the model still isn't better than random, it means something must be completely broken. If any of the experiments you did resulted in some interesting finding, please report back, since I might be able to use that information to help further.

**Comment by BenjaminBossan** [comment]
@TLampr any updates?

**Comment by BenjaminBossan** [comment]
@TLampr I added a notebook that shows how to use BERT with the help of torchtext in #645. Maybe that can help you a bit.

**Comment by BenjaminBossan** [comment]
Since there hasn't been any new comment for a while and an example is being provided, I'll close the issue for now.
